primitives/src/merkle_tree/field_based_mht/path.rs

```diff
@@ -0,0 +1,380 @@
+    #[inline]
+    pub fn is_non_empty_rightmost(&self) -> bool {
+        assert!(check_precomputed_parameters::<T>(self.path.len()));
+
+        let mut height = 0usize;
+        for &(ref siblings, direction) in &self.path {
+
+            // If the node on the path is not in the rightmost position
+            if direction != T::MERKLE_ARITY - 1 {
+
+                // If its following sibling is not the empty node, then the node
+                // cannot be the non empty rightmost at this height and for the
+                // whole tree
+                if siblings[direction] != T::EMPTY_HASH_CST.unwrap().nodes[height] {
+                    return false;
+                }
+            }
+            height += 1;
+        }
+        return true;
+    }
+
+    /// Returns the index of the leaf, corresponding to the `self` Merkle Path, in the
+    /// corresponding Merkle Tree.
+    pub fn leaf_index(&self) -> usize {
+        let mut leaf_index = 0;
+        self.path
+            .iter()
+            .enumerate()
+            .for_each(|(i, (_, pos))| leaf_index += T::MERKLE_ARITY.pow(i as u32) * pos);
+
+        leaf_index
+    }
+}
+
+impl<T: FieldBasedMerkleTreeParameters> PartialEq for FieldBasedMHTPath<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.path == other.path
+    }
+}
+
+impl<T: FieldBasedMerkleTreeParameters> FieldBasedMerkleTreePath for FieldBasedMHTPath<T> {
+    type H = T::H;
+
+    /// A Merkle Path for a leaf of a Merkle Tree with arity >= 2 will be made up of couples of nodes
+    /// and an integer. The nodes are all the siblings of the leaf ( in number MERKLE_ARITY - 1 )
+    /// and the integer is the position of the leaf, among its siblings, in the input of the hash
+    /// function (valid values are from 0 to MERKLE_ARITY - 1).
+    type Path = Vec<(Vec<<Self::H as FieldBasedHash>::Data>, usize)>;
+
+    type Parameters = T;
+
+    fn new(path: Self::Path) -> Self {
+        Self { path }
+    }
+
+    fn verify_without_length_check(
+        &self,
+        leaf: &<Self::H as FieldBasedHash>::Data,
+        root: &<Self::H as FieldBasedHash>::Data
+    ) -> Result<bool, Error> {
+
+        // Rate may also be smaller than the arity actually, but this assertion
+        // is reasonable and simplify the design. Should be also enforced by the
+        // MerkleTree that creates this instance, but let's do it again.
+        assert_eq!(<<Self::H as FieldBasedHash>::Parameters as FieldBasedHashParameters>::R, T::MERKLE_ARITY);
+
+        let mut digest = <Self::H as FieldBasedHash>::init_constant_length(T::MERKLE_ARITY, None);
+        let mut prev_node = leaf.clone();
+        for (sibling_nodes, position) in self.path.as_slice() {
+            assert_eq!(sibling_nodes.len(), T::MERKLE_ARITY - 1);
+
+            // Get the position of the node among its siblings
+            let prev_node_position = *position % T::MERKLE_ARITY;
```

Comment:
we can assume position is < MERKLE_ARITY. assert it

Reply:
Done